alphastats/tl/differential_expression_analysis.py

```diff
@@ -0,0 +1,431 @@
+    It should be subclassed for specific methods, such as t-tests or ANOVA. The class provides methods
+    for input validation, output validation, and running the analysis. It also provides a method for
+    getting the significance of the results based on a q-value cutoff.
+
+    Attributes
+    ----------
+    mat : pd.DataFrame
+        The input data for the analysis.
+    result : pd.DataFrame
+        The result of the analysis.
+
+    Methods
+    -------
+    perform(**kwargs)
+        Performs the analysis and stores the result. This fixes the workflow of input validation,
+        running the test, and validating the output.
+    get_significance(result: pd.DataFrame, qvalue_cutoff: float) -> pd.DataFrame
+        Returns the significant features based on the q-value cutoff.
+
+    Abstract Methods
+    ----------------
+    _extend_validation(**kwargs)
+        Extends the validation of parameters for the specific method.
+    _perform(**kwargs) -> pd.DataFrame
+        Wrapper that runs the statistical test from kwargs.
+    _run_statistical_test(mat: pd.DataFrame, **kwargs) -> pd.DataFrame
+        Runs the statistical test. The wrapper and the actual method are separated to allow for easier
+        testing and to ensure that all parameters are defined with types and defaults.
+
+    Intended Usage
+    --------------
+        Implement shared methods for two-group analysis.
+        Implement t-test specific methods.
+    result = dea.perform(**settings)  # run once
+    significance = dea.get_significance(cached_result, 0.05)  # run multiple times
+    volcano_plot(cached_result, significance)  # visualize
+    """
+
+    def __init__(self, mat: pd.DataFrame) -> None:
+        """Constructor for the DifferentialExpressionAnalysis class. sets up mat and results.
+
+        Parameters:
+        mat (pd.DataFrame): The input data for the analysis. This should be a DataFrame with the samples as rows and the features as columns.
+        """
+        if mat.empty:
+            raise ValueError(
+                "Input matrix to differential expression analysis is empty."
+            )
+        self.mat = mat
+        self.result: pd.DataFrame = None
+
+    def _validate_input(self, **kwargs) -> None:
+        """Abstract method to validate the input and parameters. This should raise an exception if the input or parameters are invalid
+
+        This function here checks for all parameters required for analysis regardless of the specific method, namely log2_transformed and metadata.
+
+        Parameters:
+        **kwargs (dict): The parameters for the analysis. The keys need to be defined within the allowed_parameters method.
+        """
+        try:
+            self._extend_validation(**kwargs)
+        except TypeError as err:
+            raise TypeError(
+                f"{str(err)}. Accepted keyword arguments to perform are {', '.join(self._allowed_parameters())}."
+            ) from err
+
```

Comment:
is it still needed to catch a TypeError here?

Reply:
Nope, indeed not.